llvm/lib/Target/AIE/AIEHazardRecognizer.cpp

```diff
@@ -563,3 +607,23 @@ AIEHazardRecognizer::getSelectedAltOpcode(MachineInstr *MI) const {
     return It->second;
   return std::nullopt;
 }
+
+MemoryBankBits AIEHazardRecognizer::getMemoryBanks(MachineInstr *MI) const {
+  MemoryBankBits MemoryBankUsed = 0;
+
+  if (!(MI->mayLoad() || MI->mayStore()))
+    return MemoryBankUsed;
+
+  const AIEBaseSubtarget &STI = AIEBaseSubtarget::get(*MI->getMF());
+  for (auto &MMO : MI->memoperands()) {
+    unsigned MemoryBank =
+        (STI.getMemoryBanksFromAddressSpace(MMO->getAddrSpace())).to_ulong();
+    MemoryBankUsed |= MemoryBank;
+  }
+  if (MI->mayStore() && MemoryBankUsed == 1) {
+    // If its a Store instruction and it does not have Bank info then we allow
+    // it to go with a load instruction
+    MemoryBankUsed = 0;
```

Comment:
What does the magic literal '1' mean here? And is there a good reason to treat store bank conflicts different from load bank conflicts? It looks as if supplying more information will lead to more scheduling restrictions here.

Reply:
Also, do we really need a magic bit to signify `unknown`? Why not just have `~0`, i.e. all the banks?

I would also slightly change the rationale for special-casing stores. It's mainly to maintain the status-quo, but also because in the kernels we deal with, the input and output buffers are usually in separate banks. So it is generally fine to be optimistic. Once the kernels are more specific with their bank usage, we can drop that special-casing and be fully-pessimistic when it comes to "unknown banks"

Reply:
As noted [here ](https://github.com/Xilinx/llvm-aie/pull/95/commits/a4baa6d9da1fc16d277c324a0b2c0a3c72ecf0de#diff-7b59f2472b3801ded7e5587449e12c8791893c5dc1a3a0ae58eb9f62279b6426:~:text=//%20Bit%200%20is%20for%20MI%20with%20no%20bank%20information%2C) bit 0 represent no/unknown bank annotation.  

We need to have this information to make sure we do not allow VLDA & VLDB in event we do not have bank information available.
Also "STORE" are handled differently because we would want to allow one VLD with VST, even if we do not have a bank annotation.

Reply:
Based on todays(31st July 2024) discussion we have concluded on not to treat STORES differently.

Reply:
I checked the QoR without special consideration of STORE, there are NO regression.